File: src/main/kotlin/pt/isel/ls/server/data/dataInterfaces/models/UserData.kt
package pt.isel.ls.server.data.dataInterfaces.models

import pt.isel.ls.server.data.transactionManager.transactions.TransactionCtx
import pt.isel.ls.server.User

interface UserData {

    fun createUser(name: String, email: String, hashedPassword: String, urlAvatar: String?, ctx: TransactionCtx): Pair<Int, String>

    fun getUser(token: String, ctx: TransactionCtx): User

    fun getUser(idUser: Int, ctx: TransactionCtx): User

    fun getUsers(idBoard: Int, ctx: TransactionCtx): List<User>

    fun changeAvatar(token: String, avatar: String, ctx: TransactionCtx)

    fun login(email: String, hashedPassword: String, ctx: TransactionCtx) : String
}

File: static-content/data/userData.js
import {fetchReq} from "../utils.js";
import {BASE_URL} from "../config.js";

async function createUser(name, email, password, urlAvatar) {
    return await fetchReq(
        "user",
        "POST",
        {
            name: name,
            email: email,
            password: password,
            avatar: urlAvatar
        }
    )
}

async function getUser(token) {
    return await fetchReq("user", "GET", null, token)
}

async function getAllUsers(idBoard, token) {
    return await fetchReq(`board/${idBoard}/allUsers`, "GET", null, token)
}

async function login(email, password) {
    return await fetchReq(
        "user/login",
        "POST",
        {
            email: email,
            password: password
        }
    )
}

async function changeAvatar(imgUrl) {
    return await fetchReq("user", "PUT", {imgUrl}, sessionStorage.getItem("token"))
}

async function getUserAvatar(token) {
    const noUserIcon = "../resources/images/user-unknown-icon.png"
    if(!token) return noUserIcon
    const user = await getUser(token)
    return `${user.avatar === "null" ? noUserIcon : user.avatar}`
}

export default {
    createUser,
    getUser,
    getAllUsers,
    login,
    changeAvatar,
    getUserAvatar
}